extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/NarayanaJtaRecorder.java

```diff
@@ -23,7 +25,8 @@ public void setNodeName(final TransactionManagerConfiguration transactions) {
 
         try {
             arjPropertyManager.getCoreEnvironmentBean().setNodeIdentifier(transactions.nodeName);
-            TxControl.setXANodeName(transactions.xaNodeName.orElse(transactions.nodeName));
+            jtaPropertyManager.getJTAEnvironmentBean().setXaRecoveryNodes(Collections.singletonList(transactions.nodeName));
+            TxControl.setXANodeName(transactions.nodeName);
```

Comment:
Do we genuinely need the TxControl setting here? I would imagine it can be controlled through the arjPropertyManager directly?

Reply:
this is the problematic issue of the whole narayana configuration. Once the `TxControl` class is first  initialized (and that's decision of the JVM, not possible to be influence. normally the `TxControl` is initialized as class when accessed first time) then any change done in the `arjPropertyManager` has no effect for the further use.
As we can't be sure what setter is started first or where the `TxControl` data is needed for the first time during Quarkus startup I consider this as the safe way (or maybe even necessary).

A bit related point is that I created a jira about rethinking the configuration of the whole Narayana. That came from discussion with @mmusgrov, here https://issues.jboss.org/browse/JBTM-3215

Reply:
What we do have to be certain of is there are no transactions that could be created before this is called and indeed the recovery manager has not started running before it. Are you sure that is the case?

Reply:
I assume that this code must run before the transaction service is created? In which case I would expect no transactions could be created before this runs? Or is that not the case in Quarkus? IOW could there be a transaction created with the default (non-unique) node ID and that could therefore lead to data integrity issue?

/cc @mmusgrov

Reply:
@tomjenkinson the need of use `TxControl.setXANodeName` is not caused by time when the transaction service is created. The need is caused by the fact that we can't be sure that nobody else calls the `TxControl` `setter`. When it happens then all data are initialized. When this method is called after the `setter` was used before then changes in the `bean` have no effect.

In summary, by me the `TxControl` call is needed here.

Reply:
Thanks for confirming. I suppose you are less concerned about the setter being called (which as you say it is [public](https://github.com/jbosstm/narayana/blob/master/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TxControl.java#L168)) but rather when the class is loaded it will read the value as set in the arjPropertyManager: https://github.com/jbosstm/narayana/blob/master/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TxControl.java#L238

In that case, I understand the point - thanks!